experimental/casemapping/src/provider/data.rs

```diff
@@ -0,0 +1,475 @@
+/// }
+///     4   case-sensitive
+///     3   exception
+///     2   case-ignorable
+///  1..0   0 uncased
+///         1 lowercase
+///         2 uppercase
+///         3 titlecase
+///         The runtime code relies on the case-ignorable and case type bits 2..0
+///         to be the lowest bits with this encoding.
+///
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[repr(transparent)]
+pub struct CaseMappingDataULE(RawBytesULE<2>);
+
+impl CaseMappingDataULE {
+    // 1..0 case type
+    const CASE_TYPE_BITS: u16 = 0x3;
+    // 2 case-ignorable
+    const CASE_IGNOREABLE_BIT: u16 = 0x4;
+    // 3 exception
+    const EXCEPTION_BIT: u16 = 0x8;
+    // 4 case-sensitive
+    const CASE_SENSITIVE_BIT: u16 = 0x10;
+    // 15..4 unsigned exception index
+    const EXCEPTION_SHIFT: u16 = 4;
+    // 15..7 signed-delta to simple case mapping code point (or reserved)
+    const DELTA_SHIFT: u16 = 7;
+    // 6..5 dot type
+    const DOT_TYPE_BITS: u16 = 0x60;
+    const DOT_SHIFT: u16 = 5;
+}
+
+/// # Safety
+///
+/// Safety checklist for `ULE`:
+///
+/// 1. The type *must not* include any uninitialized or padding bytes: repr(transparent)
+///    wrapper around ULE type
+/// 2. The type must have an alignment of 1 byte: repr(transparent) wrapper around ULE type
+/// 3. The impl of [`ULE::validate_byte_slice()`] *must* return an error if the given byte slice
+///    would not represent a valid slice of this type: It does
+/// 4. The impl of [`ULE::validate_byte_slice()`] *must* return an error if the given byte slice
+///    cannot be used in its entirety (if its length is not a multiple of `size_of::<Self>()`):
+///    it does, due to the RawBytesULE parse call
+/// 5. All other methods *must* be left with their default impl, or else implemented according to
+///    their respective safety guidelines: They have been
+/// 6. The equality invariant is satisfied
+unsafe impl ULE for CaseMappingDataULE {
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), ZeroVecError> {
+        let sixteens = RawBytesULE::<2>::parse_byte_slice(bytes)?;
+
+        for sixteen in sixteens {
+            let sixteen = sixteen.as_unsigned_int();
+            // The type has reserved bits in the
+            // uncased + not exception case
+            if sixteen & Self::EXCEPTION_BIT == 0 {
+                // not an exception
+                if sixteen & Self::CASE_TYPE_BITS == 0 {
+                    // uncased
+                    if sixteen >> Self::DELTA_SHIFT != 0 {
+                        // We have some used bits in the reserved zone!
+                        return Err(ZeroVecError::parse::<Self>());
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl AsULE for CaseMappingData {
+    type ULE = CaseMappingDataULE;
+
+    fn from_unaligned(ule: Self::ULE) -> Self {
+        let sixteen = ule.0.as_unsigned_int();
+
+        let ignoreable = (sixteen & CaseMappingDataULE::CASE_IGNOREABLE_BIT) != 0;
```

Comment:
Question: previously, all of these values were lazily computed. Now you're essentially pre-computing them. Does this matter? (Are there cases where you compute more things than you need?)

Reply:
Potentially. Overall I don't think it matters all that much; and we can optimize with `get_ule()` + specific methods in the future if we need for code that is in tight loops.

The main difference is that this essentially takes a lot of checks that used to be `debug_assert!()`s and sticks them in the type system, which means that check always happens but is cheap.